### PR TITLE
use ocap discipline in fixup_packages.py etc. WIP

### PR DESCRIPTION
**Pull request by dckc**
It's great to see cross-pollination between cloudabi and the python community. cc @moreati

I have made [object capability discipline in python code](http://www.madmode.com/2013/python-capability-idioms-part-1.html) a habit for the last few years. While the cloudabi-ports build tools are perhaps not a prime application for the technology, perhaps it's a good place to introduce people to the discipline.

I thought this would take just an hour or two, but of course it's taking longer, so I'll share it as-is. If you like the idea, I'll try to finish it up and make the commits look nicer.

### CODE REVIEW

**Comment by EdSchouten** [comment]
I just looked through your changes briefly, as I hope to look into them in more detail this evening. That said, overall these changes look very beautiful. Though I'm a big fan of Python, I am by no means an expert, so I truly appreciate it when people with more knowledge than me are looking into improving things.

So for now, just assume that you've got thumbs up from my side. I'll (try to) give feedback later today. 👍

File: src/config.py [hunk 1]
```diff
@@ -19,21 +18,39 @@
...
-HOST_CC = ('/usr/bin/clang-3.7' if platform.system() == 'Linux' else
-           '/usr/bin/cc')
-HOST_CXX = ('/usr/bin/clang++-3.7' if platform.system() == 'Linux' else
-            '/usr/bin/c++')
-
-# Name of the Perl executable.
-PERL = ('/usr/local/bin/perl' if platform.system() == 'FreeBSD' else
-        '/usr/bin/perl')
+
+class Platform(object):
```
Comment (EdSchouten) [inline src/config.py:30]: If we're going in this direction, would it make sense to use some kind of class hierarchy here, so we'll have a `FreeBSDPlatform` that inherits from this, but overrides `perl()`?

File: src/distfile.py [hunk 2]
```diff
@@ -101,50 +110,53 @@ def _fetch(self):
...
-            fout.write(bytes('#define _CLOUDLIBC_UNSAFE_STRING_FUNCTIONS\n',
-                             encoding='ASCII'))
-            fout.write(fin.read())
-            os.rename(path + '.new', path)
+            path = target / filename
+            with path.open('rb') as fin, (path + '.new').open('wb') as fout:
+                fout.write(
+                    bytes('#define _CLOUDLIBC_UNSAFE_STRING_FUNCTIONS\n',
+                          encoding='ASCII'))
+                fout.write(fin.read())
```
Comment (EdSchouten) [inline src/distfile.py:138]: Now that I'm reading my own code: should we replace this with `shutil.copyfileobj(fin, fout)`?

**Review by EdSchouten — COMMENTED** [review]
Very nice work so far!

By the time you're finished with this, I'd love to get your thoughts on the `BUILD` file format as well. Maybe we could improve the syntax of those as well. The `__div__()` overloading for path concatenation looks a lot nicer than the `.path()` member function we have right now.

File: src/config.py (hunk 1 above)
Comment (dckc) [reply src/config.py:30]: Perhaps; I haven't really thought that one through. I was just in ocap-refactor mode, not thinking much. I didn't start to understand what the code really does until much later.

Perhaps I should explain the technique a bit more:
1. Open a python file in emacs with flycheck and flake8
2. Review module imports; strike any that are not [powerless](http://waterken.sourceforge.net/javadoc/org/joe_e/Powerless.html) such as `import os`, `import sys`, platform (these will be imported in `_script` later)
3. C-!-n (flycheck-next-error) to visit each `undefined name` error and salt to taste
   - Add arguments to pass capabilities explicitly
     - add doctests (or static type declarations using [py2scala](https://bitbucket.org/DanC/py2scala) or mypy) to gain confidence that the refactoring is correct (_I haven't started on this part._)
4. As noted in [Capability idioms for python](http://www.madmode.com/2013/python-capability-idioms-part-1.html), if there is any code not in a function, put it in a `main` function and add call it from the top-level script environment: `if __name__ == '__main__'`
   - encapsulate powerful imports in `_script()`
     - Note that the `_initial_caps()` idiom in that 2013 blog item isn't quite right; it leaves the powerful capabilities un-encapsulated

File: src/distfile.py (hunk 2 above)
Comment (dckc) [reply src/distfile.py:138]: I don't have strong feelings one way or the other.

**Comment by dckc** [comment]
Given my struggles with integration testing, I'm trying the static typing approach with mypy. It's going well; see my [static_types_mypy branch](https://github.com/dckc/cloudabi-ports/commits/static_types_mypy)

7947a875fef0c400a002dfae3863e027e6842ad5 thru 92865423f7d724d50b2ad82cf9eade7bbc50c231

**Pull request closed without merging**